### ISSUE TITLE
In chat example, two users can kick each other

### ISSUE THREAD

**Issue by shiraeeshi** [simonmar/parconc-examples#29]
In [chat example](https://www.oreilly.com/library/view/parallel-and-concurrent/9781449335939/ch12.html#sec_chat) from Chapter 12, it seems to me that there is no guarantee that when two users kick each other at the same time, only one of the kicks will succeed.
The kick operation is divided to two STM transactions (two `atomically` calls): one of them is in `server` function in `runClient` (at [line 214](https://github.com/simonmar/parconc-examples/blob/59a4ca80d548a11d465b274a06d191e8cd6185f4/chat.hs#L214)), the second one is in `handleMessage` (at [line 236](https://github.com/simonmar/parconc-examples/blob/59a4ca80d548a11d465b274a06d191e8cd6185f4/chat.hs#L236)). When we divide the transaction into two, we lose the atomicity, right?
Here is an example project (https://github.com/shiraeeshi/hs-parconc-chat-server) that shows how to make two concurrent kicks succeed. I added `/pause`, `/resume`, `/pauseKicks`, `/resumeKicks` commands. You can see the scenario in readme.

**Comment by shiraeeshi** [comment]
You can see the changes in [this commit diff](https://github.com/shiraeeshi/hs-parconc-chat-server/commit/bbd6361af5d1b91accdd02dbe4c560cfffed01af).

**Comment by simonmar** [comment]
Yes, I think you're right. To fix this, either the two transactions should be combined, or the `kick` function should check that the current user has not already been kicked.
